internal/kafka/test/integration/kafka_create_test.go

```diff
@@ -0,0 +1,528 @@
+						mockkafkas.With(mockkafkas.CLOUD_PROVIDER, kafkaRequestPayload.CloudProvider),
+						mockkafkas.With(mockkafkas.REGION, kafkaRequestPayload.Region),
+						mockkafkas.With(mockkafkas.OWNER, account2.Username()),
+						mockkafkas.With(mockkafkas.ORGANISATION_ID, account2.Organization().ExternalID()),
+						mockkafkas.With(mockkafkas.INSTANCE_TYPE, types.STANDARD.String()),
+						mockkafkas.With(mockkafkas.SIZE_ID, "x1"),
+					))
+				}
+
+				db := h.DBFactory().New()
+				err := db.Create(&dummyKafkas).Error
+				g.Expect(err).NotTo(gomega.HaveOccurred(), "failed to create dummy kafkas")
+			},
+			assert: func() {
+				_, resp, err := common.WaitForKafkaCreateToBeAccepted(ctx, kafkatest.TestServices.DBFactory, client, kafkaRequestPayload)
+				if resp != nil {
+					resp.Body.Close()
+				}
+				g.Expect(err).NotTo(gomega.HaveOccurred(), "Error posting object:  %v", err)
+				// get a 403 status code as there is no capacity left
+				g.Expect(resp.StatusCode).To(gomega.Equal(http.StatusForbidden), "kafka should have been rejected as no capacity is left")
+			},
+			cleanup: func() {
+				// delete dummy kafkas to free up resources on the OSD cluster
+				db := h.DBFactory().DB.Model(&dbapi.KafkaRequest{})
+				var kafkas []*dbapi.KafkaRequest
+				err := db.Scan(&kafkas).Error
+				g.Expect(err).ToNot(gomega.HaveOccurred(), "failed to get all kafka in the database")
+
+				ctx := h.NewAuthenticatedContext(account2, nil)
+
+				for _, kafka := range kafkas {
+					deleteTestKafka(t, h, ctx, client, kafka.ID)
+				}
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		testcase := tc
+		t.Run(testcase.name, func(t *testing.T) {
+			if testcase.setup != nil {
+				testcase.setup()
+			}
+
+			testcase.assert()
+
+			if testcase.cleanup != nil {
+				testcase.cleanup()
+			}
+		})
+	}
+}
+
+// Test Kafka creation with dynamic scaling enabled
+func TestKafkaCreate_DynamicScaling(t *testing.T) {
+	g := gomega.NewWithT(t)
+
+	ocmServer := mocks.NewMockConfigurableServerBuilder().Build()
+	defer ocmServer.Close()
+
+	var enableAutoscale bool
+	h, client, teardown := kafkatest.NewKafkaHelperWithHooks(t, ocmServer, func(d *config.DataplaneClusterConfig) {
+		if enableAutoscale {
+			d.DataPlaneClusterScalingType = config.AutoScaling
+			d.DynamicScalingConfig = config.DynamicScalingConfig{
+				Configuration: map[string]config.InstanceTypeDynamicScalingConfig{
+					"standard": {
+						ReservedStreamingUnits: 0,
+					},
+					"developer": {
+						ReservedStreamingUnits: 0,
+					},
+				},
+			}
+		}
+	})
+	defer teardown()
+
+	// set up data plane cluster
+	// The mock fleetshard sync only updates the cluster if it's in a non-ready state. Only wait until the cluster has status
+	// waiting_for_kas_fleetshard_operator to ensure that the mock fleetshard sync updates the dynamic capacity info correctly.
+	clusterID, getClusterErr := common.GetOSDClusterIDAndWaitForStatus(h, t, api.ClusterWaitingForKasFleetShardOperator)
```

Comment:
I don't think this is coupled with the cluster_mgr_test. If there wasn't any cluster available, this GetOSDCluster should create a new one.

Alternative is to create a cluster, with a `waiting_for_fleetshard_operator` status, by manually creating it with db client. It would save time by bypassing cluster reconcile up to this point. Would this be preferred?

Reply:
> I don't think this is coupled with the cluster_mgr_test. If there wasn't any cluster available, this GetOSDCluster should create a new one.

Where I see the coupling is that this method call will wait for the cluster to be almost fully terrraformed thus following the reconciliation logic similar to cluster_mgr_test. This method call will then either reuses the cluster from the cluster_mgr_test or spins up a new one and starts to terraform it until it reaches that stataus. 

> Alternative is to create a cluster, with a waiting_for_fleetshard_operator status, by manually creating it with db client. It would save time by bypassing cluster reconcile up to this point. Would this be preferred?

+1. The terraforming process takes a bit of time and thus my preference of directly seeding the cluster in the database with the given status. To ensure that we won't have any issues with the seeded cluster not being available in OCM, the seeded cluster ProviderType has to be `standalone`.

Reply:
sounds good :+1: